hnn_core/neuron.py

```diff
@@ -0,0 +1,539 @@
+
+    if rank == 0:
+        for spike_vec in spiketimes_list:
+            neuron_net._all_spiketimes.append(spike_vec)
+        for spike_vec in spikegids_list:
+            neuron_net._all_spikegids.append(spike_vec)
+
+    _PC.barrier()  # get all nodes to this place before continuing
+
+    dpl_data = np.c_[np.array(dp_rec_L2.to_python()) +
+                     np.array(dp_rec_L5.to_python()),
+                     np.array(dp_rec_L2.to_python()),
+                     np.array(dp_rec_L5.to_python())]
+
+    dpl = Dipole(np.array(t_vec.to_python()), dpl_data)
+    if rank == 0:
+        if neuron_net.net.params['save_dpl']:
+            dpl.write('rawdpl.txt')
+
+        dpl.baseline_renormalize(neuron_net.net.params)
+        dpl.convert_fAm_to_nAm()
+        dpl.scale(neuron_net.net.params['dipole_scalefctr'])
+        dpl.smooth(neuron_net.net.params['dipole_smooth_win'] / h.dt)
+
+    neuron_net.net.trial_idx += 1
+
+    return dpl
+
+
+def load_custom_mechanisms():
+    import platform
+    import os.path as op
+
+    global _LOADED_DLL
+
+    if _LOADED_DLL is not None:
+        return
+
+    if platform.system() == 'Windows':
+        mech_fname = op.join(op.dirname(__file__), '..', 'mod', 'nrnmech.dll')
+    else:
+        mech_fname = op.join(op.dirname(__file__), '..', 'mod', 'x86_64',
+                             '.libs', 'libnrnmech.so')
+    h.nrn_load_dll(mech_fname)
+    _LOADED_DLL = mech_fname
+
+    if _get_rank() == 0:
+        print('Loading custom mechanism files from %s' % mech_fname)
+
+    return
+
+
+def _get_nhosts():
+    """Return the number of processors used by ParallelContext
+
+    Returns
+    -------
+    nhosts: int
+        Value from pc.nhost()
+    """
+    if _PC is not None:
+        return int(_PC.nhost())
+
+    return 1
+
+
+def _get_rank():
+    """Return the MPI rank from ParallelContext
+
+    Returns
+    -------
+    rank: int
+        Value from pc.id()
+    """
+    if _PC is not None:
+        return int(_PC.id())
+
+    return 0
+
+
+def _create_parallel_context(n_cores=None):
+    """Create parallel context.
+
+    Parameters
+    ----------
+    n_cores: int | None
+        Number of processors to use for a simulation. A value of None will
+        allow NEURON to use all available processors.
+    """
+
+    global _CVODE, _PC
+
+    if _PC is None:
+        if n_cores is None:
+            # MPI: Initialize the ParallelContext class
+            _PC = h.ParallelContext()
+        else:
+            _PC = h.ParallelContext(n_cores)
+
+        _CVODE = h.CVode()
+
+        # be explicit about using fixed step integration
+        _CVODE.active(0)
+
+        # use cache_efficient mode for allocating elements in contiguous order
+        # cvode.cache_efficient(1)
+    else:
+        # ParallelContext() has already been called. Don't start more workers.
+        # Just tell old nrniv workers to quit.
+        _PC.done()
+
+
+class NeuronNetwork(object):
+    """The NeuronNetwork class.
```

Comment:
so I'm just trying to wrap my head around this class. The idea is that this class contains everything in the old Network class that pertains to Neuron software? The advantage is that the new Network class is picklable?

One disadvantage I see with this approach is that it disallows users to pass any Neuron objects that they created. I can see that becoming useful if we wanted to allow more detailed cell morphologies or dynamics for instance. See [my proposal](https://github.com/jonescompneurolab/hnn-core/issues/111#issuecomment-657740682) for incorporating the calcium dynamics for instance. Not sure if that's the best approach but I can see those kinds of APIs being useful in similar scenarios that involve customization.

Reply:
Yes, and:
1. network.py is simpler; no reason for context manager there
2. `NeuronNetwork` can be rebuilt. For running batches of multiple trials on the same network.

Then parallel simulation would break.

Reply:
Edit to 2) above. Really for running multiple trials without respawning nrniv. Running multiple trials on the same network is a different idea that I don't think is implemented yet.

Reply:
humm okay, can we document this somewhere? Maybe as a new section called Notes in the docstring? See [how we do it in MNE](https://mne.tools/stable/generated/mne.Epochs.html#mne.Epochs)

Reply:
Yeah, this would be helpful to document. Done. There will be more nuances to document as we wrestle with running batch simulations.